### DISCUSSION TITLE
PlaceStub Filter

### DISCUSSION THREAD

**Discussion by dkdeshAcademic** [RosettaCommons/rosetta#104] [RosettaScripts]
Hi, I am trying to dock a scaffold protein onto a target protein in the presence of hotspot residues (specifically ARG and TYR).
When I run my xml protocol, I get the following summarized error:

**core.pack.pack_rotamers: built 181 rotamers at 181 positions.
core.pack.interaction_graph.interaction_graph_factory: Instantiating LinearMemoryInteractionGraph
protocols.hotspot_hashing: Associated stubs with scaffold chain 2
protocols.hotspot_hashing: Making hotspot constraints...
protocols.hotspot_hashing: Applied 169 hotspot constraints to the pose.
protocols.protein_interface_design.movers.PlaceStubMover: adding multiple stub constraints to pose
removing stub constraints from pose
constraint score evaluates to 0 in this pose. Failing
reapplied 174 bb constraints to pose
protocols.rosetta_scripts.ParsedProtocol: Mover PlaceStub reports failure!
protocols.jd2.JobDistributor: [ WARNING ] model_8_0001 reported failure 1 times and will no longer retry (permanent failure)**

I just don't know what it means when it says "constraint score evaluates to 0 in this pose" and haven't found a suitable fix. If someone could give me an answer that'd be great thanks! Below is my xml file:

<ROSETTASCRIPTS>

    <SCOREFXNS>
        <ScoreFunction name="ddg_scorefxn" weights="score12_full" hs_hash="0.0"/>
        <ScoreFunction name="stub_docking_low" weights="interchain_cen" hs_hash="10.0"/>
    </SCOREFXNS>

    <FILTERS>
        <Ddg name="ddg" threshold="-10" scorefxn="ddg_scorefxn" repack="0" confidence="1"/>
        <Sasa name="sasa" threshold="800"/>
            <ShapeComplementarity name="shape" verbose="1" min_sc="0.4" confidence="1"/>
        <CompoundStatement name="ddg_sasa"> 
               <AND filter_name="ddg"/>
               <AND filter_name="sasa"/>
           </CompoundStatement>
           <TerminusDistance name="termini" distance="0"/>
    </FILTERS>

    

    <TASKOPERATIONS>
        <RestrictToRepacking name="rtr"/>
    </TASKOPERATIONS>

    <MOVERS>
        <Docking name="dock" fullatom="0" local_refine="0" score_low="stub_docking_low"/>
			<RepackMinimize name="des1" minimize_bb="0" minimize_rb="1" scorefxn_repack="soft_rep" scorefxn_minimize="score_docking" interface_cutoff_distance="10.0"/>
			<RepackMinimize name="des2" minimize_bb="0" minimize_rb="1" interface_cutoff_distance="10.0"/>
			<RepackMinimize name="des3" minimize_bb="1" minimize_rb="1" interface_cutoff_distance="10.0"/>
			<SaveAndRetrieveSidechains name="srsc"/>
            <SetupHotspotConstraints name="hcst" stubfile="allStubs.pdb" cb_force="0.5"/>
			<PlaceStub name="place_arg" minimize_rb="1" stubfile="ARG_combinedStubs.pdb" stub_energy_threshold="10" final_filter="ddg_sasa" hurry="true" max_cb_dist="4.0">
					<DesignMovers>
						<Add mover_name="srsc"/>
						<Add mover_name="des1" coord_cst_std="1.5"/>
						<Add mover_name="des2" coord_cst_std="2"/>
						<Add mover_name="des3"/>
					</DesignMovers>
			</PlaceStub>
			<PlaceStub name="place_leu" place_scaffold="1" stubfile="LEU_combinedStubs.pdb" add_constraints="1" minimize_rb="1" hurry="true" score_threshold="10.0" stub_energy_threshold="3.0" max_cb_dist="4.0">
					<StubMinimize>
                        <Add mover_name="place_arg" bb_cst_weight="0.4"/>
                    </StubMinimize>
                    <DesignMovers>
						<Add mover_name="place_arg" coord_cst_std="1.5" use_constraints="true"/>
					</DesignMovers>
			</PlaceStub>
    </MOVERS>

    
    <PROTOCOLS>
        <Add mover_name="dock" filter_name="termini"/>
        <Add mover_name="hcst"/>
          <Add mover_name="place_leu"/>
          <Add filter_name="ddg"/>
          <Add filter_name="sasa"/>
</PROTOCOLS>
    
</ROSETTASCRIPTS>

**Comment by roccomoretti** [comment]
PlaceStub uses constraints to keep things in place. If these constraints aren't fulfilled (the constraint score evaluates to 0), then it's a bad stub placement and things early-out.

Given that you're passing an explicit stubfile to the PlaceStub mover, I would check your stub definitions and make sure they're valid. If there's something off with them, it could be that the constraints being applied are invalid and thus aren't being scored properly.

**Reply by dkdeshAcademic** [reply]
I've made sure the stubs are on the structure I want, and saw this error message while I was running it:

And now minimizing all dofs together
protocols.protein_interface_design.movers.PlaceStubMover: backbone_stub_constraint score changed from -5.49714 to 0 during minimization
removing hotspot constraints from pose
DEBUG.PlaceStubMover: removed 114 bb constraints from pose
before removal 114 after removal 0
Csts before min 0 after min 0
protocols.protein_interface_design.movers.PlaceStubMover: removed all hotspot constraints
bb stub constraint score too high. Skipping this stub
constraint score evaluates to 0 in this pose. Failing
reapplied 114 bb constraints to pose
protocols.rosetta_scripts.ParsedProtocol: Mover PlaceStub reports failure!
protocols.jd2.JobDistributor: [ WARNING ] model_1_0001 reported failure 1 times and will no longer retry (permanent failure)

**Reply by roccomoretti** [reply]
`backbone_stub_constraint score changed from -5.49714 to 0 during minimization` is the core part of your issue. You started with a reasonable (a negative) stub constraint score, but during minimization of your stub, something caused it to move from the position it should be in, to somewhere where the constraints no longer apply, resulting in a 0 constraint energy.

That's not unexpected. That's why there's a backbone constraint filter in the first place. The stochastic nature of most Rosetta protocols means that often you'll run a large number of potential outputs, and only a fraction of them will be decent. Here we're just filtering them before 

I'd recommend re-running the protocol with a larger `-nstruct` value. Don't worry too much if you don't get all the structures you asked for, so long as you're getting some. It's only if you're getting all (or almost all) of the potential outputs fail should you worry that something might be off.

Note that if you want `-nstruct` to be the actual number of structures that are output, you can alternatively set `-ntrials`. -ntrials will set how many times a failed job (or at least one which failed due to filtering) will be re-run in an attempt to pass the filters. If you set it reasonably large (e.g. 100), when Rosetta encounters a failure it will restart the job until it gets a usable output, or it tries -ntrials times. This won't fix consistent issues leading to filtering failures, but it may help if you're anticipating missing a few structures due to stochastic issues.

**Reply by dkdeshAcademic** [reply]
Yeah, the protocol conducts 99 jobs and every single one of the fails due to the same issue. Is this normal?

**Reply by dkdeshAcademic** [reply]
I keep getting that "constraint score evaluates to 0 in this pose. Failing" error

**Reply by dkdeshAcademic** [reply]
The way I created my stub files was copy pasting the coordinates of each stub into a file and segmenting which stub was which with model_001, etc. Do you have a more efficient way of creating stub files?
